src/repr/src/adt/timestamp.rs

```diff
@@ -704,6 +795,52 @@ impl<T: TimestampLike> CheckedTimestamp<T> {
         // If at any point we overflow, map to a TimestampError.
         age_inner(self, other).ok_or(TimestampError::OutOfRange)
     }
+
+    /// Rounds the timestamp to the specified number of digits of precision.
+    pub fn round_to_precision(&mut self, precision: Option<TimestampPrecision>) {
+        let precision = precision.map(|p| p.into_u8()).unwrap_or(MAX_PRECISION);
+        // maximum precision is micros
+        let power = MAX_PRECISION
+            .checked_sub(precision)
+            .expect("precision fits in micros");
+        let round_to_micros = 10_i64.pow(power.into());
+
+        let mut original = self.date_time();
+        let nanoseconds = original.timestamp_subsec_nanos();
+        // truncating to microseconds does not round it up
+        // i.e. 123456789 will be truncated to 123456
+        original = original.truncate_microseconds();
+        // depending upon the 7th digit here, we'll be
+        // adding 1 millisecond
+        // so eventually 123456789 will be rounded to 123457
+        let seventh_digit = (nanoseconds % 1_000) / 100;
+        assert!(seventh_digit < 10);
+        if seventh_digit >= 5 {
+            original = original + Duration::microseconds(1);
+        }
+        // this is copied from [`chrono::round::duration_round`]
+        // but using microseconds instead of nanoseconds precision
+        let stamp = original.timestamp_micros();
+        let dt = {
+            let delta_down = stamp % round_to_micros;
+            if delta_down == 0 {
+                original
+            } else {
+                let (delta_up, delta_down) = if delta_down < 0 {
+                    (delta_down.abs(), round_to_micros - delta_down.abs())
+                } else {
+                    (round_to_micros - delta_down, delta_down)
+                };
+                if delta_up <= delta_down {
+                    original + Duration::microseconds(delta_up)
+                } else {
+                    original - Duration::microseconds(delta_down)
+                }
+            }
+        };
+
+        self.t = T::from_date_time(dt)
```

Comment:
```suggestion
        Self::from_timestamplike(dt)
```

Recommendation to change this fn to return a new CheckedTimestamp instead of mutate self, which we don't do anywhere else. This will also ensure that should the round fn can never produce an out-of-range timestamp.

Reply:
done!